File: python/mirage/commands/builtin/generic/gzip.py
import zlib
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass

from mirage.commands.builtin.utils.constants import STDIN_OPERAND
from mirage.commands.builtin.utils.stream import (operand_label,
                                                  resolve_source, stdin_bytes)
from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.constants import flag_kwarg_name
from mirage.commands.spec.flag_view import FlagView
from mirage.commands.spec.types import FlagValue
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec
from mirage.utils.compress import gunzip_checked, gzip_compress_stream
from mirage.utils.errors import GzipDataError
from mirage.utils.key_prefix import mounted_path


def extract_level(fl: FlagView) -> int:
    """The compression level -1..-9 asked for, or zlib's default.

    The digits are short-only options, so each is its own dest -- except
    ``-1``, which the parser disambiguates to ``args_1``
    (``AMBIGUOUS_NAMES``). Reading the bag by the bare digit therefore
    missed ``gzip -1`` entirely.

    Args:
        fl (FlagView): Flag view constructed with the gzip spec.
    """
    for n in range(9, 0, -1):
        if fl.as_bool(flag_kwarg_name(str(n))):
            return n
    return zlib.Z_DEFAULT_COMPRESSION


async def gzip(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    unlink: Callable[..., Awaitable[None]],
    stdin: ByteSource | None = None,
    decompress: bool = False,
    keep: bool = False,
    force: bool = False,
    to_stdout: bool = False,
    level: int = zlib.Z_DEFAULT_COMPRESSION,
) -> tuple[ByteSource | None, IOResult]:
    if not paths and not decompress:
        return gzip_compress_stream(resolve_source(stdin),
                                    level=level), IOResult()
    read = stdin_bytes(read_bytes, stdin)
    writes: dict[str, ByteSource] = {}
    stdout: list[bytes] = []
    errors: list[str] = []
    # With no operand gzip -d reads stdin. A `-` has no file to replace, so
    # it goes to stdout; gzip refuses to follow /dev/stdin in place, so
    # only `-` does this. An input with no gzip header is reported and
    # skipped; a truncated or corrupt one ends the run.
    for p in paths or [STDIN_OPERAND]:
        in_place = not (to_stdout or p.raw_path == "-")
        raw = await (read_bytes(p) if in_place else read(p))
        if decompress:
            try:
                data = gunzip_checked(raw)
            except GzipDataError as exc:
                errors.append(f"gzip: {operand_label(p, 'stdin')}: {exc}\n")
                if exc.fatal:
                    break
                continue
        else:
            data = zlib.compress(raw, level=level, wbits=zlib.MAX_WBITS | 16)
        if not in_place:
            stdout.append(data)
            continue
        stripped = p.mount_path
        if decompress:
            out_path = stripped.removesuffix(".gz") if stripped.endswith(
                ".gz") else stripped + ".out"
        else:
            out_path = stripped + ".gz"
        await write_bytes(mounted_path(p, out_path), data)
        writes[out_path] = data
        if not keep:
            await unlink(p)
    return b"".join(stdout) or None, IOResult(writes=writes,
                                              exit_code=1 if errors else 0,
                                              stderr="".join(errors).encode()
                                              or None)


__all__ = ["gzip", "extract_level"]


@dataclass(frozen=True, slots=True)
class GzipFlags:
    decompress: bool = False
    keep: bool = False
    force: bool = False
    to_stdout: bool = False
    level: int | None = None


def parse_flags(flags: Mapping[str, FlagValue]) -> GzipFlags:
    fl = FlagView(flags, spec=SPECS["gzip"])
    return GzipFlags(
        decompress=fl.as_bool("d"),
        keep=fl.as_bool("k"),
        force=fl.as_bool("f"),
        to_stdout=fl.as_bool("c"),
        level=extract_level(fl),
    )


def gzip_writes(flags: Mapping[str, FlagValue], paths: list[PathSpec]) -> bool:
    """Whether a gzip invocation writes: each file operand is replaced by
    its archive unless ``-c`` sends the result to stdout, while a ``-``
    operand, like no operand, filters stdin to stdout.

    Args:
        flags (Mapping[str, FlagValue]): the parsed flag bag.
        paths (list[PathSpec]): the operands the mount received.
    """
    replaces = any(p.raw_path != "-" for p in paths)
    return replaces and not parse_flags(flags).to_stdout


async def gzip_generic(
    paths: list[PathSpec],
    texts: list[str],
    opts: CommandOpts,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    unlink: Callable[..., Awaitable[None]],
) -> tuple[ByteSource | None, IOResult]:
    parsed = parse_flags(opts.flags)
    return await gzip(paths,
                      read_bytes=read_bytes,
                      write_bytes=write_bytes,
                      unlink=unlink,
                      stdin=opts.stdin,
                      decompress=parsed.decompress,
                      keep=parsed.keep,
                      force=parsed.force,
                      to_stdout=parsed.to_stdout,
                      level=(parsed.level if parsed.level is not None else
                             zlib.Z_DEFAULT_COMPRESSION))

File: typescript/packages/core/src/commands/builtin/generic/gzip.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { specOf } from '../../spec/builtins.ts'
import { FlagView } from '../../spec/flag_view.ts'
import { mountedPath } from '../../../utils/key_prefix.ts'
import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import { gunzipChecked, gzip } from '../../../utils/compress.ts'
import { GzipDataError } from '../../../utils/errors.ts'
import type { CommandFnResult, CommandOpts, WritesFn } from '../../config.ts'
import { STDIN_OPERAND } from '../utils/constants.ts'
import { operandLabel, resolveSource, stdinStream } from '../utils/stream.ts'

const ENC = new TextEncoder()

function concat(chunks: Uint8Array[]): Uint8Array {
  let total = 0
  for (const c of chunks) total += c.byteLength
  const out = new Uint8Array(total)
  let offset = 0
  for (const c of chunks) {
    out.set(c, offset)
    offset += c.byteLength
  }
  return out
}

// Whether a gzip invocation writes: each file operand is replaced by its
// archive unless -c sends the result to stdout, while a `-` operand, like no
// operand, filters stdin to stdout. Mirrors Python's gzip_writes.
export const gzipWrites: WritesFn = (flags, paths) =>
  paths.some((p) => p.rawPath !== '-') && !new FlagView(flags, specOf('gzip')).asBool('c')

export async function gzipGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
  write: (p: PathSpec, data: Uint8Array) => Promise<void>,
  unlink: (p: PathSpec) => Promise<void>,
): Promise<CommandFnResult> {
  const fl = new FlagView(opts.flags, specOf('gzip'))
  const decompress = fl.asBool('d')
  const keep = fl.asBool('k')
  const stdoutMode = fl.asBool('c')

  if (paths.length === 0 && !decompress) {
    const result: ByteSource = await gzip(await materialize(resolveSource(opts.stdin)))
    return [result, new IOResult()]
  }
  const read = stdinStream(stream, opts.stdin)
  const writes: Record<string, Uint8Array> = {}
  const stdout: Uint8Array[] = []
  let errors = ''
  // With no operand gzip -d reads stdin. A `-` has no file to replace, so it
  // goes to stdout; gzip refuses to follow /dev/stdin in place, so only `-`
  // does this. An input with no gzip header is reported and skipped; a
  // truncated or corrupt one ends the run.
  for (const p of paths.length > 0 ? paths : [STDIN_OPERAND]) {
    const inPlace = !(stdoutMode || p.rawPath === '-')
    const raw = await materialize(inPlace ? stream(p) : read(p))
    let data: Uint8Array
    if (decompress) {
      try {
        data = await gunzipChecked(raw)
      } catch (err) {
        if (!(err instanceof GzipDataError)) throw err
        errors += `gzip: ${operandLabel(p, 'stdin')}: ${err.message}\n`
        if (err.fatal) break
        continue
      }
    } else {
      data = await gzip(raw)
    }
    if (!inPlace) {
      stdout.push(data)
      continue
    }
    const pStripped = p.mountPath
    let outPath: string
    if (decompress) {
      outPath = pStripped.endsWith('.gz') ? pStripped.slice(0, -3) : pStripped + '.out'
    } else {
      outPath = pStripped + '.gz'
    }
    await write(mountedPath(p, outPath), data)
    writes[outPath] = data
    if (!keep) await unlink(p)
  }
  return [
    stdout.length > 0 ? concat(stdout) : null,
    new IOResult({
      writes,
      exitCode: errors === '' ? 0 : 1,
      stderr: errors === '' ? null : ENC.encode(errors),
    }),
  ]
}
